supabase/migrations/14_gateway_auth.sql

```diff
@@ -48,29 +48,38 @@ returns text as $$
 
 $$ language sql stable security definer;
 
-
-
 create function gateway_auth_token(variadic prefixes text[])
 returns table (token text, gateway_url text) as $$
-
-  with authorized_prefixes as (
-    select p as prefix
-    from auth_roles() as r, unnest(prefixes) as p
-    where starts_with(p, r.role_prefix)
-  )
-
-  select internal.sign_jwt(
+declare
+  requested_prefixes int := (select count(distinct p) from unnest(prefixes) p);
+  authorized_prefixes text[];
+begin
+  
+  select array_agg(distinct p) into authorized_prefixes
+    from 
+      unnest(prefixes) as p
+      join auth_roles() as r on starts_with(p, r.role_prefix);
+
+  -- authorized_prefixes will be null when _none_ of the requested prefixes are authorized.
+  -- In that case the array_length comparison won't work, so we need an explicit null check.
+  if authorized_prefixes is null or array_length(authorized_prefixes, 1) != requested_prefixes then
```

Comment:
I'm not entirely positive I'm reading this right, but is this checking array length instead of array contents? Like, if you're authorized for X but request Y, then this won't throw so long as the number of requests and authorizations match?

Reply:
Yeah, we're only checking the length of the arrays here. This works because `authorized_prefixes` is assigned by doing an inner join to the auth roles, so there's no way for it to contain prefixes that aren't a part of `prefixes`.